### DISCUSSION TITLE
Completion stops working intermittently with lsp-mode

### DISCUSSION THREAD

**Discussion by lassemaatta** [minad/corfu#374] [Q&A] [answered]
# Problem

Sometimes completing a symbol with `corfu` gets stuck returning "No match", even though everything worked fine a moment ago. 

# Setup 

I use `corfu` with `lsp-mode` as the LSP client and `clojure-lsp` as the LSP server. Emacs is compiled from the 29 branch (`GNU Emacs 29.1.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-09-30`). My emacs [config](https://github.com/lassemaatta/dotfiles/blob/master/.emacs.d/settings.org), which should hopefully follow the `corfu`+`lsp-mode` instructions from the wiki.

Disclaimer: I'm very much a emacs-beginner, so it's more than likely that this is just some user error or problem in my emacs configuration.

# How to reproduce and/or resolve

When testing this I tried to completions on the following Clojure form `(http/|)`, where `http` is an symbolic alias for a HTTP library namespace and `|` represents the cursor position. The expected result is seeing something like:

![Screenshot_20230930_095638](https://github.com/minad/corfu/assets/18479268/50a1489d-1fdd-4677-9a74-716e0e28a302)

Unfortunately I don't have a fool-proof way to reproduce this nor how to fix it once it happens. 

It feels like there's some kind of caching issue. For example, if completion for some form (e.g. `(http/|)`) gets stuck returning "No match" then attempting completion for another form might work fine. Sometimes doing stuff like changing the alias (e.g. rename `http` to `https` and then completing on `(https/|)`) or modifying the file and saving it seems to fix the problem. At least for a while. 

I tried playing around with the `cape-capf-buster` in a very "I have no idea what I'm doing" sort of way, but it did not appear to affect the situation.

# Debugging

At first I looked at the communication between the LSP client and the server. I have no idea how the LSP protocol and the completion system in emacs interact, but from what I've deducted: something (`corfu`, `lsp-mode`?) will fetch completions from LSP server and cache them. That is, when everything is working fine attempting completion may or may not ask completions from the LSP server. 

To debug `corfu`, I used the `cape-capf-debug` thing to wrap `lsp-completion-at-point`.

<details>
 <summary><b>Log - Completion success</b></summary>
<pre>
lsp-completion-at-point@126 => input=223:228:"http/" table=(#("http/*current-middleware*" 0 25 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/*current-middleware*" 0 25 ...) "kind" 6 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "*current-middleware*" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 1119 "name-col" 3))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/accept-encoding-value" 0 26 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/accept-encoding-value" 0 26 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "accept-encoding-value" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 732 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/basic-auth-value" 0 21 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/basic-auth-value" 0 21 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "basic-auth-value" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 821 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/check-url!" 0 15 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/check-url!" 0 15 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "check-url!" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 1163 "name-col" 12))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/client-error?" 0 18 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/client-error?" 0 18 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "client-error?" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 224 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) ...) annotation-function=lsp-completion--annotate company-kind=lsp-completion--candidate-kind company-deprecated=lsp-completion--candidate-deprecated company-require-match=never company-prefix-length=t company-match=lsp-completion--company-match company-doc-buffer=#<function> exit-function=#<function>
lsp-completion-at-point@126(action=metadata input=223:228:"http/" prefix="http/" ignore-case=nil) => (metadata (category . lsp-capf) (display-sort-function . identity) (cycle-sort-function . identity))
lsp-completion-at-point@126(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil predicate=#<function>) => nil
lsp-completion-at-point@126(action=metadata input=223:228:"http/" prefix="http/" ignore-case=nil) => (metadata (category . lsp-capf) (display-sort-function . identity) (cycle-sort-function . identity))
lsp-completion-at-point@126(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil) => nil [2 times]
lsp-completion-at-point@126(action=all input=223:228:"http/" prefix="" ignore-case=t regexp=("\\(?:http/\\)")) => (#("http/*current-middleware*" 0 25 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/*current-middleware*" 0 25 ...) "kind" 6 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "*current-middleware*" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 1119 "name-col" 3))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/accept-encoding-value" 0 26 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/accept-encoding-value" 0 26 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "accept-encoding-value" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 732 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/basic-auth-value" 0 21 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/basic-auth-value" 0 21 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "basic-auth-value" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 821 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/check-url!" 0 15 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/check-url!" 0 15 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "check-url!" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 1163 "name-col" 12))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) #("http/client-error?" 0 18 (lsp-completion-item #s(hash-table size 6 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("http/client-error?" 0 18 ...) "kind" 3 "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("unresolved" [["documentation" #s(hash-table size 4 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("name" "client-error?" "uri" "file:///home/lassemaatta/.emacs.d/workspace/.cache/clj_http.client.clj" "name-row" 224 "name-col" 7))] ["alias" #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("ns-to-add" "clj-http.client" "alias-to-add" "http" "uri" "file:///home/lassemaatta/work/uimaranta_basis/src/uimaranta_basis/core.clj"))]])) "score" 8 "_emacsStartPoint" 223)) lsp-sort-text nil lsp-completion-start-point 223 lsp-completion-markers ... lsp-completion-prefix "http/" match-data ...)) ...)
lsp-completion-at-point@126(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil) => nil
</pre>
</details>

<details>
 <summary><b>Log - Completion failure</b></summary>
<pre>
lsp-completion-at-point@93 => input=223:228:"http/" table=nil annotation-function=lsp-completion--annotate company-kind=lsp-completion--candidate-kind company-deprecated=lsp-completion--candidate-deprecated company-require-match=never company-prefix-length=t company-match=lsp-completion--company-match company-doc-buffer=#<function> exit-function=#<function>
lsp-completion-at-point@93(action=metadata input=223:228:"http/" prefix="http/" ignore-case=nil) => (metadata (category . lsp-capf) (display-sort-function . identity) (cycle-sort-function . identity))
lsp-completion-at-point@93(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil predicate=#<function>) => nil
lsp-completion-at-point@93(action=all input=223:228:"http/" prefix="" ignore-case=t regexp=("\\(?:http/\\)") predicate=#<function>) => nil
lsp-completion-at-point@93(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil) => nil
lsp-completion-at-point@93(action=try input=223:228:"http/" prefix="http/" ignore-case=nil) => nil
lsp-completion-at-point@93(action=(boundaries . "") input=223:228:"http/" prefix="http/" ignore-case=nil) => nil
lsp-completion-at-point@93(action=all input=223:228:"http/" prefix="http" ignore-case=nil regexp=("\\`http.*?/")) => nil
</pre>
</details>

Comparing the two I can see that the success case has an entry with a bunch of results:
- `(action=all input=223:228:"http/" prefix="" ignore-case=t regexp=("\\(?:http/\\)"))`
while the problematic case as a similarish looking row with an extra predicate (no idea if this is relevant)
- `(action=all input=223:228:"http/" prefix="" ignore-case=t regexp=("\\(?:http/\\)") predicate=#)`

**Comment by minad** [comment]
Do you use https://github.com/minad/corfu/wiki#configuring-corfu-for-lsp-clients?

**Reply by lassemaatta** [reply]
I believe so. The relevant parts of my config:

```elisp
(use-package corfu
  :init
  (global-corfu-mode)
  (corfu-popupinfo-mode)
  (corfu-history-mode -1)
  (add-to-list 'savehist-additional-variables 'corfu-history)
  :custom
  (corfu-auto nil)
  (corfu-preselect 'prompt)
  (corfu-popupinfo-delay '(0.5 . 1.0))
  (corfu-cycle t)
  (corfu-on-exact-match nil)
  (corfu-quit-no-match nil)
  (corfu-preview-current nil)
  (corfu-min-width 70)
  (corfu-count 20))

(use-package orderless
  :demand t
  :after vertico
  :custom
  (completion-styles '(basic orderless partial-completion))
  (completion-category-defaults nil)
  (completion-category-overrides '((file (styles basic partial-completion)))))

(use-package cape
  :demand t
  :after vertico)

(use-package lsp-mode
  :custom
  (lsp-completion-provider :none)
  (lsp-completion-sort-initial-results nil)
  ;; set prefix for lsp-command-keymap (few alternatives - "C-l", "C-c l")
  (lsp-keymap-prefix "C-f l")
  (lsp-idle-delay 0.5)
  (lsp-eldoc-enable-hover t)
  (lsp-lens-enable t)
  (lsp-headerline-breadcrumb-enable nil)
  ;; Close LSP server when last buffer closed
  (lsp-keep-workspace-alive nil)
  ;; Don't include the var/function declaration itself when listing references to a var/function
  (lsp-references-exclude-definition t)
  ;; We use yasnippet, but not through LSP
  (lsp-enable-snippet nil)
  ;; Prefer clojure-mode indentation
  (lsp-enable-indentation nil)
  (lsp-modeline-code-actions-enable nil)
  (lsp-imenu-index-function #'lsp-imenu-create-categorized-index)
  ;; For logging IO between client and server
  (lsp-log-io nil)
  ;(lsp-clojure-custom-server-command '("bash" "-c" "/home/lassemaatta/Lataukset/clojure-lsp"))
  :init
  (defun my/lsp-mode-setup-completion ()
    (setf (alist-get 'styles (alist-get 'lsp-capf completion-category-defaults))
          '(orderless)))
  :hook (;; replace XXX-mode with concrete major-mode(e. g. python-mode)
         (clojure-mode . lsp)
         (clojurescript-mode . lsp)
         (clojurec-mode . lsp)
         (js-mode . lsp)
         (typescript-mode . lsp)
         ;; if you want which-key integration
         (lsp-mode . lsp-enable-which-key-integration)
         (lsp-completion-mode . my/lsp-mode-setup-completion)
         (lsp-after-apply-edits . (lambda (op)
                                    (save-some-buffers t))))
  :bind (("<C-return>" . lsp-find-definition)
         ("<C-.>" . lsp-find-definition)
         ("<C-M-return>" . lsp-describe-thing-at-point))
  :commands (lsp lsp-deferred))
```

**Reply by lassemaatta** [reply, marked as answer]
I did a bit more testing. I disabled `corfu` and used the default completion method (not sure what it's called, the `*Completions*` buffer thing) and the same problem still occurs. Also, starting a "plain LSP" instance with `lsp-start-plain` also exhibits this issue. Therefore, I think I'll raise this issue over at `lsp-mode` and see if they have any ideas what's going on.

**Reply by minad** [reply]
Thanks. Yes, reporting this to lsp-mode with a self-contained recipe would be good.

**Reply by emagieika** [reply]
Hey @lassemaatta, I've just encountered something similar and in my case this was`(corfu-quit-no-match nil)` that I accidentally enabled and after typing something like a single letter, lsp wouldn't return a match, because of minimum length required and after I finished typing what I needed "No Match" persisted.
Not sure if this is expected or not (from the variable name it kind of sounds like it is), but leaving `(corfu-quit-no-match t)` solved the problem for me.
